Create recipe section

Klaypexx opened the pull request.

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

gitignor don't check image foldr

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

Klaypexx committed:
commit

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Types/types.ts:
```diff
@@ -28,10 +28,18 @@ interface AddRecipeButtonProps {
-  margin?: boolean;
   className?: string;
-  children?: React.ReactNode;
+  props?: {
```
Советую весь этот файл разбить на отдельные файлы, т.е. 1 интерфейс - 1 файл, я не скажу что это общепринятая практика везде, но у нас так
Плюс будет тебе респект за улучшение читаемости кода и проекта вообще
(эту правку лучше оформить отдельным пуллреквестом)

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Services/RecipeService.ts:
```diff
@@ -0,0 +1,44 @@
+};
+
+const createRecipe = async (recipeData: FormData) => {
+  const response = await api.post(endpoints.create, recipeData, {
```
хочется видеть больше типизации. т.е. что за response, что в нем есть, скорее всего это какие-то типы из либы, надо только их заиспользовать

k-sannikov replied:
Это касается не только этого момента, а всего проекта
Пробегись по проекту, посмотри, где у тебя есть нетипизированные моменты

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Pages/RecipesList/RecipesList.tsx:
```diff
@@ -0,0 +1,66 @@
+    const fetchRecipes = async () => {
+      try {
+        const response: RecipeListValues[] = await RecipeService.getAllRecipes(page);
+        if (response.length < 4) {
```
вынеси 4 в константу

Klaypexx replied:
пока удалил цифру, потом подумаю еще над этой проверкой

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Components/Form/RecipeForm/RecipeValidation.ts:
```diff
@@ -0,0 +1,37 @@
+      Yup.object().shape({
+        description: Yup.string()
+          .required('Описание шага обязательно')
+          .max(300, 'Описание шага не должно превышать 300 символов'),
```
300, 20 и другие числа желательно выносить в константы, чтобы потом, если придется менять их, не забыть поменять их во всех местах

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Components/Form/LoginForm/LoginForm.module.css:
он нужен?

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Components/Field/TagsField/TagsField.tsx:
```diff
@@ -1,23 +1,32 @@
 
-      if (tags.length >= 5) {
-        setError('Вы не можете добавить больше 5 тегов'); // Установка ошибки
+      if (inputValue.length > 20) {
```
20 в константу

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Components/Field/CardField/CardField.tsx:
```diff
@@ -1,50 +1,87 @@
+                <option value="30">30</option>
+                <option value="35">35</option>
+                <option value="40">40</option>
+                <option value="60">60</option>
```
как-то бы в цикле генерить эти options

k-sannikov commented on CookingRecipesApplication/CookingRecipes/src/Components/Field/CardField/CardField.tsx:
```diff
@@ -1,50 +1,87 @@
+                <option value="2">2</option>
+                <option value="3">3</option>
+                <option value="4">4</option>
+                <option value="5">5</option>
```
как-то бы в цикле генерить эти options

k-sannikov commented on CookingRecipesApplication/CookingRecipes/.vite/deps/package.json:
такое ощущуение. что это должно быть в гитигноре?

k-sannikov commented on CookingRecipesApplication/CookingRecipes/.vite/deps/_metadata.json:
такое ощущуение. что это должно быть в гитигноре?

k-sannikov commented on CookingRecipesApi/Infrastructure/Tags/Repository/TagRepository.cs:
```diff
@@ -0,0 +1,37 @@
+            return null;
+        }
+
+        var tags = await _entitiesTag
```
var на явные типы заменить

k-sannikov commented on CookingRecipesApi/Infrastructure/Tags/Repository/TagRepository.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        _entitiesTag = context.Set<Tag>();
+    }
+    public async Task<List<RecipeTag>> GetOrCreateTag( List<string>? tagNames )
```
нарушается принцип единой ответственности, у тебя метод должен делать что-то одно, а ты в него поместил 2 абсолютно разные функции

k-sannikov replied:
Очень много логики в репозитории. так не надо. репозиторий должен быть очень простой, что-то что-то получить. что-то удалить, что-то создать
Обычно метод создания один, метод удаления один, а методов получения какой-тоо инфы несколько, т.к. приходится делать выборку по разным условиям в разных ситуациях

k-sannikov commented on CookingRecipesApi/Infrastructure/Recipes/Repositories/RecipeRepository.cs:
```diff
@@ -0,0 +1,45 @@
+         .ThenInclude( recipeTag => recipeTag.Tag )
+         .Include( recipe => recipe.Author )
+         .OrderBy( recipe => recipe.Id )
+         .Skip( ( page - 1 ) * 4 )
```
вот этот параметр "( page - 1 ) * 4"
и вот этот "4"
надо передавать в метод репозитория, а не хардкодить и вычислять их внутри

k-sannikov replied:
также название не соответствует логике метода, надо переименовать
оно в любом случае будет длинное, но зато понятное

k-sannikov commented on CookingRecipesApi/Infrastructure/Recipes/Repositories/RecipeRepository.cs:
```diff
@@ -0,0 +1,45 @@
+         .ToListAsync();
+    }
+
+    public async Task<Recipe> GetCurrentUserRecipe( int recipeId )
```
название не соответствует функционалу, тут будто бы должно быть что-то типа "GetByIdWithTagsAndIngredientsAndStepsAndAuthor", но такое название уж слишком длинное, если ты просто инклюдишь вообще все что можно (тут надо еще подумать. нужно ли оно тебе прямо всё), то можно назвать "GetByIdWithAllDetails"

k-sannikov replied:
форматирование поправь, тут отступы неправильные

k-sannikov replied:
в этом репозитории после с сущностями называется _entities. в другом как-то иначе, приди к единому стилю (я за _entities)

k-sannikov commented on CookingRecipesApi/Infrastructure/Recipes/Repositories/RecipeRepository.cs:
```diff
@@ -0,0 +1,45 @@
+         .Include( recipe => recipe.Ingredients )
+         .Include( recipe => recipe.Steps )
+         .Include( recipe => recipe.Author )
+         .OrderBy( recipe => recipe.Id )
```
OrderBy точно нужен?

Klaypexx replied:
Раньше без него у меня появлялось предупреждение. Сейчас когда тестил этого предупреждения в консоли уже не было.

k-sannikov commented on CookingRecipesApi/Infrastructure.Migrations/Migrations/AppDbContextModelSnapshot.cs:
```diff
@@ -304,9 +298,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
-                        .HasForeignKey("RecipeId")
-                        .OnDelete(DeleteBehavior.Cascade)
-                        .IsRequired();
+                        .HasForeignKey("RecipeId");
```
круто что смог избавиться от OnDelete(DeleteBehavior.Cascade)

k-sannikov commented on CookingRecipesApi/Infrastructure.Migrations/Migrations/20240801122322_Init2.cs:
```diff
@@ -0,0 +1,22 @@
+namespace Infrastructure.Migrations.Migrations
+{
+    /// <inheritdoc />
+    public partial class Init2 : Migration
```
миграции мы называем оснознанно, как и коммиты (хотя с названием коммитов у тебя немного беда🤨)

k-sannikov replied:
еще и миграция пустая. очень странно, надо бы разобраться

k-sannikov commented on CookingRecipesApi/Infrastructure.Migrations/Migrations/20240801113913_Init.cs:
```diff
@@ -49,7 +49,7 @@ protected override void Up(MigrationBuilder migrationBuilder)
                     name = table.Column<string>(type: "nvarchar(max)", nullable: false),
                     description = table.Column<string>(type: "nvarchar(max)", nullable: false),
-                    time = table.Column<TimeOnly>(type: "time", nullable: false),
+                    time = table.Column<int>(type: "int", nullable: false),
```
очень сильно не советую вручную править миграции
Если у тебя есть с ними проблема, то могу посоветовать просто их все удалить и заново сгенерить (перед этим не забудь в базе все почистить)

k-sannikov replied:
но в реальных проектах так не работает, т.к. тут как с гитом, лучше не ошибаться, замучаешься разгребать проблемы

Klaypexx replied:
Не думаю что я сам это правил, создавал 2 миграции Init и Init2 кажется. В миграции сам никогда не лезу.

k-sannikov commented on CookingRecipesApi/Domain/Recipes/Entities/Recipe.cs:
```diff
@@ -3,16 +3,15 @@
     public int AuthorId { get; set; }
     public List<Like> Likes { get; set; }
-    public List<RecipeTag> Tags { get; set; }
+    public List<RecipeTag>? Tags { get; set; }
```
зачем сделал нуллбл свойства? они же и так могут быть нулл, без этого обозначения, да еще и не понятно почему, например, Tags нуллбл, а Likes нет. надо бы к единообразию привести

k-sannikov commented on CookingRecipesApi/Domain/Recipes/Entities/Recipe.cs:
```diff
@@ -3,16 +3,15 @@
-    public string Avatar { get; set; }
+    public int CookingTime { get; set; }
+    public int Portion { get; set; }
+    public string? Avatar { get; set; }
```
если тут нуллбл чисто для того, чтобы правильно миграция создалась. то лучше нуллбл тип определить в конфигурации этой сущности, а не тут

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Program.cs:
```diff
@@ -85,6 +99,7 @@
 app.UseHttpsRedirection();
 
 app.UseCors( "MyPolicy" );
+app.UseStaticFiles();
```
прикольно, не знал про эту штуку

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/Validators/RecipeDtoValidator.cs:
```diff
@@ -0,0 +1,78 @@
+    public TagDtoValidator()
+    {
+        RuleFor( s => s.Name )
+            .MaximumLength( 20 ).WithMessage( "Тег не должен превышать 300 символов" );
```
опечатка, чтобы их избежать лучше эти числа вынести в константы

Klaypexx replied:
Где их лучше было бы хранить?

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/Validators/RecipeDtoValidator.cs:
```diff
@@ -0,0 +1,78 @@
+            .SetValidator( new IngredientDtoValidator() );
+    }
+
+    private bool BeAnImage( IFormFile file )
```
обычно методы которые возвращают bool принято называть Is..., Has..., тут советую назвать IsImage()

k-sannikov replied:
вынести бы этот метод в какой-то класс хелпер. типа холпер для работы с файлами

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/Validators/RecipeDtoValidator.cs:
```diff
@@ -0,0 +1,78 @@
+
+    private bool BeAnImage( IFormFile file )
+    {
+        if ( file == null )
```
если файл нулл, то это изображение???

Klaypexx replied:
Если true, то он не выведет ошибку, так как avatar может быть пустым.

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/RecipesDto/StepDto.cs:
```diff
@@ -0,0 +1,8 @@
+namespace CookingRecipesApi.Dto.RecipesDto;
+public class StepDto
+{
+    [Required]
```
[Required] для чего ты добавил? Для сваггера?

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/RecipesDto/RecipeDto.cs:
```diff
@@ -0,0 +1,20 @@
+    public string Description { get; set; }
+    [Required]
+    public int CookingTime { get; set; }
+    [Required]
```
[Required] для чего ты добавил? Для сваггера?

k-sannikov replied:
если что эти комменты сразу по всем ДТО

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Dto/RecipesDto/RecipeDto.cs:
```diff
@@ -0,0 +1,20 @@
+    public int CookingTime { get; set; }
+    [Required]
+    public int Portion { get; set; }
+    public IFormFile? Avatar { get; set; }
```
если ? для сваггера, то норм, если нет, то напиши для чего или убери лучше

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/CookingRecipesApi.csproj:
```diff
@@ -27,4 +28,8 @@
     <ProjectReference Include="..\Infrastructure\Infrastructure.csproj" />
   </ItemGroup>
 
+	<ItemGroup>
```
отступы

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+    [HttpPost]
+    [Route( "get/{userId}" )]
+    [Authorize]
+    public async Task<IActionResult> GetCurrentUserRecipe( [FromBody] int recipeId )
```
ничего не понимаю, userId в роуте, recipeId FromBody

k-sannikov replied:
еще и получение какой-то информации методом post сделано

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+    [HttpPost]
+    [Route( "get" )]
+    [Authorize]
+    public async Task<IActionResult> GetAllRecipes( [FromBody] int page = 1 )
```
тоже получение информации, а метод post

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+        try
+        {
+            Recipe recipes = await _recipeService.GetCurrentUserRecipe( recipeId );
+            var recipeDto = recipes.ToCurrentRecipeDto();
```
var

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+        }
+        catch ( Exception exception )
+        {
+            return BadRequest( new { message = exception.Message } );
```
мы договаривались, что ты сделаешь отдельный класс для ошибок

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+
+            int authorId = int.Parse( User.GetUserId() );
+
+            List<RecipeTag> Tags = await _tagService.GetOrCreateTag( recipeDto.Tags?.Select( tagDto => tagDto.Name ).ToList() );
```
орматирование (название переменной) + длинная строка

k-sannikov replied:
про GetOrCreateTag уже писал

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+
+        try
+        {
+            if ( recipeDto.Avatar != null )
```
все что внутри этого if вынести в сервис, оператор using можно сократить до одной строки

k-sannikov replied:
еще название фото тебе советую генерировать в виде guid, тогда не будет проблем коллизии названий

k-sannikov replied:
и не нужна будет авторизация на получение статики
т.к. гуид подобрать нереально

k-sannikov commented on CookingRecipesApi/CookingRecipesApi/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,116 @@
+        IWebHostEnvironment appEnvironment,
+        IValidator<RecipeDto> recipeDtoValidator )
+    {
+        _unitOfWork = unitOfWork;
```
_unitOfWork не используется тут, но если это все работает, значит ты куда-то на уровень ниже его засунул, надо найти это место и убрать оттуда его

k-sannikov commented on CookingRecipesApi/Application/Tags/Services/TagService.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        _tagRepository = tagRepository;
+    }
+    public async Task<List<RecipeTag>> GetOrCreateTag( List<string>? tagNames )
```
не должно быть такого метода, ищи другие комментарии по поводу него

k-sannikov commented on CookingRecipesApi/Application/Tags/Services/ITagService.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Application.Tags.Services;
+public interface ITagService
+{
+    Task<List<RecipeTag>> GetOrCreateTag( List<string>? tagNames );
```
уже отписал по этому методу

k-sannikov commented on CookingRecipesApi/Application/Tags/Repositories/ITagRepository.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Application.Tags.Repositories;
+public interface ITagRepository
+{
+    Task<List<RecipeTag>> GetOrCreateTag( List<string>? tagNames );
```
уже отписал по этому методу

k-sannikov commented on CookingRecipesApi/Application/Recipes/Services/RecipeService.cs:
```diff
@@ -0,0 +1,31 @@
+    public async Task CreateRcipe( Recipe recipe )
+    {
+        await _recipeRepository.CreateRecipe( recipe );
+        await _unitOfWork.Save();
```
Ай-яй-яй, убирает это отсюда

k-sannikov commented on CookingRecipesApi/Application/Recipes/Services/RecipeService.cs:
```diff
@@ -0,0 +1,31 @@
+        await _unitOfWork.Save();
+    }
+
+    public async Task<List<Recipe>> GetAllRecipes( int page )
```
pageNumber

k-sannikov replied:
ну и не GetAll получается

Klaypexx committed:
Рефакторинг типов

Разбил типы из файла types на файлы для каждого типа.
Исправил пути для каждого компонента.

Klaypexx committed:
Удалил тип и компонент

Удалил тип и компонент за ненадобностью

Klaypexx committed:
Валидация response

Отвалидировал response
Привел запросы и результат к единому виду

Klaypexx committed:
Добавил константы

Добавил файлы констант участвующих в валидации логина, регистрации и создания рецепта.

Klaypexx committed:
Обновлен gitignore

Обновил гитигнор, теперь он не должен читать папку .vite

Klaypexx committed:
Обновлен gitignore

Обновил gitignore

Klaypexx committed:
Обновлен gitignore

Klaypexx committed:
Рефактор TagRepository

Рефактор репозитория под единую ответственность

Klaypexx committed:
Хардкодинг в Skip

Вынес вычисления из skip

Klaypexx committed:
Исправление названия и единообразие entities

Klaypexx committed:
убрал OrderBy

Klaypexx committed:
Поправил nullable, привел к единообразию

Klaypexx committed:
Поправил опечатку

Klaypexx committed:
Поправил валидацию для Avatar

Klaypexx committed:
Убрал [Required] для Dto свойств

Klaypexx committed:
Отсутпы

Klaypexx committed:
Исправил post запросы на get

Klaypexx committed:
исправил var на конкретный тип

Klaypexx committed:
исправил форматирование

Klaypexx committed:
Исправил название изображения на guid. Перенес проверку isImage

Klaypexx committed:
Сделал отдельный класс для ошибок

Klaypexx committed:
Удалил unitOfWork с сервиса

Klaypexx committed:
Переименовал метод получение рецептов для страницы рецептов

Klaypexx committed:
Добавил константы для валидации

Klaypexx committed:
Исправление валидации

Вынес проверку на уникальность из файла валидатора в контроллер. Добавил сервис IsUniqueUsername

Klaypexx committed:
Исправил запрос c FromHeader на FromRoute

Klaypexx commented on CookingRecipesApi/CookingRecipesApi/Controllers/AuthController.cs:
```diff
@@ -76,21 +84,21 @@ public async Task<IActionResult> Login( [FromBody] LoginDto body )
+            return BadRequest( new ErrorResponse( "Неверный пароль" ) );
         }
 
         string jwtToken = _tokenService.GenerateJwtToken( user );
```
Выделить 3 строчки в authService. Назвать SignIn

Klaypexx committed:
Создание SignIn, мелкие правки

Klaypexx committed:
рефакторинг метода CreateRecipe

Klaypexx committed:
добавил try catch в auth

Klaypexx committed:
фикс

Klaypexx committed:
Удаление рецепта

Klaypexx committed:
Рефакторинг удаления рецепта

Klaypexx committed:
Рефакторинг удаления рецепта

Klaypexx committed:
Начало создание редактирования рецепта

Klaypexx committed:
Добавление функционала редактирования рецепта

Klaypexx committed:
Исправление метода удаления рецепта

Klaypexx committed:
Добавление функционала редактирования рецепта

Klaypexx committed:
Редактирование рецепта

Klaypexx committed:
Редактирование рецепта

Почистил код

Klaypexx committed:
Сохранение

Klaypexx committed:
Рефакторинг редактирования рецепта

Klaypexx committed:
Чистка кода

Klaypexx committed:
Чистка кода

Klaypexx committed:
Обновление

Добавил спиннеры для загрузки
Почистил код
Пофиксил баг с созданием рецепта

Klaypexx committed:
Merge branch 'ReactSpinners' into CreateRecipe
